build-tools/packages/build-cli/src/commands/generate/typetests.ts

```diff
@@ -243,16 +246,13 @@ export function getTypesPathFromPackage(
 			conditions: ["types"],
 			require: true,
 		});
-		// Only assign typesPath if it wasn't already assigned earlier.
-		typesPath ??=
-			exports === undefined || exports.length === 0
-				? packageJson.types ?? packageJson.typings
-				: exports[0];
+		typesPath = exports === undefined || exports.length === 0 ? undefined : exports[0];
 	} catch {
 		// Catch any exceptions here; we'll return undefined instead of throwing them.
 	}
 
-	return typesPath;
+	// Fall back to types/typings fields if both import and require conditions yielded nothing.
+	return typesPath ?? packageJson.types ?? packageJson.typings;
```

Comment:
I now see a problem. ;)
This needs to follow https://www.typescriptlang.org/docs/handbook/modules/reference.html#packagejson-main-and-types. There are two things missing:
1. This should not apply if the package has "exports" or there is a level (subpath).
2. If neither of those is present, then "main" should be used with ".d.ts" extension substitute.
...assuming you are going for really generic. The other "missing" thing is to find and @types/ matching package. But I think we are okay to constrain to our well-typed packages.

Reply:
Issue 1 was causing other downstream problems and should be fixed in #20989. I suggest holding off on the other logic unless it proves necessary (doubtful since we're pretty good about having a types field in our packages).